chore(rust): ockam command add subcommand to generate man pages

hargut opened the pull request:
<!-- Thank you for sending a pull request :heart: -->

## Current Behavior

Currently there are no man pages available for ockam.

## Proposed Changes

The PR adds a subcommand called `manpages` which generates man pages and writes the files to a user selected directory, the users default man page location within the home directory or to a fallback location based on the current working directory.

One man page is generated for the main command, and one for each subcommand. For further details please refer to the according issue:
https://github.com/build-trust/ockam/issues/3808

## Checks

<!-- To help us review and merge this pull request quickly, please confirm the following:  -->

- [x] All commits in this Pull Request are [signed](https://docs.github.com/en/authentication/managing-commit-signature-verification/signing-commits) and Verified by Github.
- [x] All commits in this Pull Request follow the Ockam [commit message convention](https://github.com/build-trust/.github/blob/main/CONTRIBUTING.md#commit-messages).
- [x] I accept the Ockam Community [Code of Conduct](https://github.com/build-trust/.github/blob/main/CODE_OF_CONDUCT.md).
- [x] I have accepted the Ockam [Contributor License Agreement](https://github.com/build-trust/ockam-contributors/blob/main/CLA.md) by adding my Git/Github details in a row at the end of the [CONTRIBUTORS.csv](https://github.com/build-trust/ockam-contributors/blob/main/CONTRIBUTORS.csv) file in a separate pull request to the [build-trust/ockam-contributors](https://github.com/build-trust/ockam-contributors) repository.

<!-- Looking forward to merging your contribution!! -->

Commit:
chore(rust): ockam command add subcommand to generate man pages

adrianbenavides commented on implementations/rust/ockam/ockam_command/src/manpages.rs:
```diff
@@ -0,0 +1,134 @@
+    }
+
+    // generate man page for sub commands
+    for s_cmd in cmd.get_subcommands() {
```
I think this loop should be run recursively, as most (if not all) of the subcommands have 1 or 2 layers of nested subcommands.

hargut replied:
Good input, thank you.
I've have switched that part to a recursive call.

adrianbenavides reviewed (commented):
This looks pretty awesome, thanks @hargut for taking a look into this!

I'm not sure we need a subcommand for this, but rather a simple function to generate the man files from CI. Also, we don't know yet how we want to organize the man files, but we can leave that for a future PR.

In any case, this is a great start. I left a small comment :+1:

hargut commented:
@adrianbenavides 
with the input from your comment I've done the following adaptions:
- extracted the manpage generation to a function
- switched to a recursive call for generating the man pages to cover all levels of subcommands
- added the hidden flag to the manpages subcommand so it does not appear for users in the help, but could be used e.g. for CI integration
- fixed some formatting issues

Initial my thought was to add the subcommand so the distribution process can further be based on a single binary.

adrianbenavides commented on implementations/rust/ockam/ockam_command/src/manpages.rs:
```diff
@@ -0,0 +1,140 @@
+        },
+    };
+
+    println!("man pages directory: {}", man_dir.display());
```
If the user provides a directory, it's already known. If the output is stored in `~/.local/share/man/man1/`, the user doesn't really care because it's the default directory and will be able to access them directly when running `man ockam`.

So, I'd show the directory only in the case we store the output in a non expected directory (after line 70). And I would change the message to `println!("Man pages stored at: {}", man_dir.display());`

hargut replied:
Changed.

adrianbenavides commented on implementations/rust/ockam/ockam_command/src/manpages.rs:
```diff
@@ -0,0 +1,140 @@
+
+impl ManpagesCommand {
+    pub fn run(self) {
+        write_man_pages(&self.dir, &self.no_compression);
```
Given that the `run` function only calls one function, I'd include the `write_man_pages` contents directly here to avoid a level of indirection.

hargut replied:
Inlined that part.

adrianbenavides commented on implementations/rust/ockam/ockam_command/src/manpages.rs:
```diff
@@ -0,0 +1,140 @@
+        Err(error) => panic!("Error getting man page directory: {:?}", error),
+    };
+    let clap_command = <OckamCommand as CommandFactory>::command();
+    generate_man_pages(man_dir.as_path(), &clap_command, None, no_compression);
```
You should move the `no_compression` boolean instead of passing the reference to this function. It's cheap to clone (although you don't need to in this case) and it's more comfortable to use.

hargut replied:
Done.

adrianbenavides commented on implementations/rust/ockam/ockam_command/src/manpages.rs:
```diff
@@ -0,0 +1,140 @@
+    };
+
+    println!("man pages directory: {}", man_dir.display());
+    if !man_dir.exists() {
+        create_dir(man_dir.clone())?;
+    }
```
Nit: you can replace this with `create_dir_all`, as you don't have to check if it already exists.

hargut replied:
Switched.

adrianbenavides reviewed (commented):
I left some more suggestions. And here are a couple of extra things:
- When you are done with all the changes, could you squash all the commits and force push?
- When I tried the command on my machine, the output contained colors (like `[1;4mFeedback: [0m    `). I think this is due to the `COLORFGBG` environment variable. Can you take a look at this? Maybe there is a way to unset that variable. If you don't see an easy way to do this, we can tackle it on a separate PR.

hargut commented:
Hi @adrianbenavides,l

I've addressed the inputs from the review and squashed all changes into a single commit.

Some background and information on the colour topic:
- Currently the colours are added in [help.rs](https://github.com/build-trust/ockam/blob/develop/implementations/rust/ockam/ockam_command/src/help.rs#L21) which is called during adding the help messages for clap (e.g [node/mod.rs](https://github.com/build-trust/ockam/blob/develop/implementations/rust/ockam/ockam_command/src/node/mod.rs#L124)).
- Within clap there is only the colourized string version available, which ended up in the man page.
- Without removing the syntax highlighting for colour enabled terminals within the help messages a way to solve that problem is to filter the ascii control characters before writing the man page to disk. This is what I've added to the PR.

I'm not very happy with that solution, and think it might be cleaner to think about removing the colourization within help messages.


Best regards,
Harald

adrianbenavides reviewed (approved):
Awesome! I agree that we have to revisit how we handle colorization, but for now, this solution is perfect. We'll create more issues on this topic to keep improving the CLI docs.

Again, thanks @hargut for this contribution! :pray:
